src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,367 @@
+    // read-only. Specifically only the following list-based attributes match the
+    // "global attributes not in metadata" (see GlobalAttributes.h :: GlobalAttributesNotInMetadat):
+    //   - AttributeList
+    //   - EventList
+    //   - AcceptedCommands
+    //   - GeneratedCommands
+    //
+    // Given the above, UnsupportedWrite should be correct (attempt to write to a read-only list)
+    bool isReadOnly = (attributeMetadata == nullptr) || (*attributeMetadata)->IsReadOnly();
+
+    // Internal is allowed to bypass timed writes and read-only.
+    if (!request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        VerifyOrReturnError(!isReadOnly, CHIP_IM_GLOBAL_STATUS(UnsupportedWrite));
+
+        VerifyOrReturnError(!(*attributeMetadata)->MustUseTimedWrite() ||
+                                request.writeFlags.Has(InteractionModel::WriteFlags::kTimed),
+                            CHIP_IM_GLOBAL_STATUS(NeedsTimedInteraction));
+    }
+
+    // Extra check: internal requests can bypass the read only check, however global attributes
+    // have no underlying storage, so write still cannot be done
+    VerifyOrReturnError(attributeMetadata != nullptr, CHIP_IM_GLOBAL_STATUS(UnsupportedWrite));
+
+    if (request.path.mDataVersion.HasValue())
+    {
+        std::optional<InteractionModel::ClusterInfo> clusterInfo = GetClusterInfo(request.path);
+        if (!clusterInfo.has_value())
+        {
+            ChipLogError(DataManagement, "Unable to get cluster info for Endpoint %x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_ERROR_NOT_FOUND;
+        }
+
+        if (request.path.mDataVersion.Value() != clusterInfo->dataVersion)
+        {
+            ChipLogError(DataManagement, "Write Version mismatch for Endpoint %x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+    }
+
+    AttributeAccessInterface * aai       = GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId);
+    std::optional<CHIP_ERROR> aai_result = TryWriteViaAccessInterface(request.path, aai, decoder);
+
+    if (aai_result.has_value())
+    {
+        return *aai_result;
+    }
+
+    ReturnErrorCodeIf(aai_result.has_value(), *aai_result);
+
+    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, gEmberAttributeIOBufferSpan));
+
+    EmberAfAttributeSearchRecord record;
+    record.endpoint    = request.path.mEndpointId;
+    record.clusterId   = request.path.mClusterId;
+    record.attributeId = request.path.mAttributeId;
+
+    Protocols::InteractionModel::Status status = emAfReadOrWriteAttribute(
+        &record, attributeMetadata, gEmberAttributeIOBufferSpan.data(), static_cast<uint16_t>(gEmberAttributeIOBufferSpan.size()),
```

Comment:
This looks wrong; `gEmberAttributeIOBufferSpan.size()` might be bigger than the size allocated in the attribute store for this attribute, since it's sized to the _max_ attribute size.

Reply:
From what I can see, typeSensitiveMemCopy at https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/attribute-storage.cpp#L486

sets the bufferSize to am-size if write is true.

This seems to be inline with the documentation of emAfRedorWriteAttribute at https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/attribute-storage.cpp#L568

```
// When writing attributes, readLength is ignored.  For non-string attributes,
// this function assumes the source buffer is the same size as the attribute
// type.  For strings, the function will copy as many bytes as will fit in the
// attribute.  This means the resulting string may be truncated.  The length
// byte(s) in the resulting string will reflect any truncated.
```

I will update comments here.